File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
nums = [list(map(int, list(input().strip()))) for _ in range(n)]

for i in range(1, n):
    for j in range(1, m):
        if nums[i][j] == 1:
            nums[i][j] = min(nums[i - 1][j], nums[i][j - 1], nums[i - 1][j - 1]) + 1

ans = 0
for x in nums:
    temp = max(x)
    ans = max(ans, temp)
print(ans ** 2)

File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.java
import java.io.*;
import java.util.*;

public class Main {
    public static int n, m, ans;
    public static int[][] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[n][m];
        ans = 0;

        for (int i = 0; i < n; i++) {
            String str = br.readLine();
            for (int j = 0; j < m; j++) {
                arr[i][j] = str.charAt(j) - '0';
                if (ans == 0 && arr[i][j] == 1) {
                    ans = 1;
                }
            }
        }

        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                if (arr[i][j] == 1) {
                    int temp = Math.min(arr[i - 1][j], arr[i][j - 1]);
                    arr[i][j] = Math.min(temp, arr[i - 1][j - 1]) + 1;
                    ans = Math.max(ans, arr[i][j]);
                }
            }
        }

        System.out.println(ans * ans);
    }
}